src/samples/perf-benchmarks/PerfProtocol.cs

```diff
@@ -1,87 +1,256 @@
-                bytesWritten += bytesToWrite;
-                _logger?.LogDebug($"Sent {request.Length} more bytes");
+                    TimeSeconds = elapsedSeconds,
+                    Type = "intermediary",
+                    DownloadBytes = _lastReportRead
+                };
+
+                string jsonResult = JsonConvert.SerializeObject(result);
+                Console.WriteLine(jsonResult);
+
+                _lastReportTime = now;
+                _lastReportRead = 0;
             }
-        });
+        }
 
-        long bytesRead = 0;
-        for (; ; )
+        public override long Seek(long offset, SeekOrigin origin) => _innerStream.Seek(offset, origin);
+        public override void SetLength(long value) => _innerStream.SetLength(value);
+        public override void Write(byte[] buffer, int offset, int count) => _innerStream.Write(buffer, offset, count);
+
+        protected override void Dispose(bool disposing)
         {
-            ReadOnlySequence<byte> read = await downChannel.ReadAsync(0, ReadBlockingMode.WaitAny).OrThrow();
-            _logger?.LogDebug($"DIAL READ {read.Length}");
-            bytesRead += read.Length;
-            if (bytesRead == TotalLoad)
+            if (disposing)
             {
-                _logger?.LogInformation($"DIAL DONE");
-                return;
+                _innerStream.Dispose();
             }
+            base.Dispose(disposing);
         }
     }
 
-    public async Task ListenAsync(IChannel downChannel, IChannelFactory upChannelFactory, IPeerContext context)
+    public class Result
     {
-        ulong total = await downChannel.ReadVarintUlongAsync();
-        ulong bytesRead = 0;
-        for (; ; )
+        public double TimeSeconds { get; set; }
+        public string Type { get; set; }
+        public ulong DownloadBytes { get; set; }
+        public ulong UploadBytes { get; set; }
+    }
+
+    public class PerfProtocol : IProtocol
+    {
+        private readonly ILogger<PerfProtocol> _logger;
+        public string Id => "/perf/1.0.0";
+        private PeerId peerId;
+        private readonly PerfConfig _config;
+        private Random rand = new();
+
+        public PerfProtocol(ILogger<PerfProtocol> logger, PerfConfig config)
+        {
+            _logger = logger;
+            _config = config;
+        }
+
+        public async Task DialAsync(IChannel downChannel, IChannelFactory upChannelFactory, IPeerContext context)
+        {
+            byte[] u64Buf = new byte[8];
+            using (Stream str = new ChannelStream(downChannel))
+            {
+                int bytesRead = await str.ReadAsync(u64Buf, 0, u64Buf.Length);
+                if (bytesRead != u64Buf.Length)
+                {
+                    _logger.LogError("Could not read the required number of bytes");
+                    str.Close();
+                    return;
+                }
+
+                ulong bytesToSend = BitConverter.ToUInt64(u64Buf, 0);
```

Comment:
You can use something like in `YamuxHeader.FromBytes` method, so no need in `ReverseBytes` implementation